jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala

```diff
@@ -0,0 +1,426 @@
+        instance.fromParamsToXGBParamMap("custom_obj") == null,
+        "we do not support persist XGBoostEstimator with customized evaluator and objective" +
+          " function for now")
+      implicit val format = DefaultFormats
+      implicit val sc = super.sparkSession.sparkContext
+      DefaultXGBoostParamsWriter.saveMetadata(instance, path, sc)
+    }
+  }
+
+  private class XGBoostClassifierReader extends MLReader[XGBoostClassifier] {
+
+    override def load(path: String): XGBoostClassifier = {
+      val metadata = DefaultXGBoostParamsReader.loadMetadata(path, sc)
+      val cls = Utils.classForName(metadata.className)
+      val instance =
+        cls.getConstructor(classOf[String]).newInstance(metadata.uid).asInstanceOf[Params]
+      DefaultXGBoostParamsReader.getAndSetParams(instance, metadata)
+      instance.asInstanceOf[XGBoostClassifier]
+    }
+  }
+}
+
+class XGBoostClassificationModel private[ml](
+    override val uid: String,
+    override val numClasses: Int,
+    private[spark] val _booster: Booster)
+  extends ProbabilisticClassificationModel[Vector, XGBoostClassificationModel]
+    with XGBoostClassifierParams with MLWritable with Serializable {
+
+  import XGBoostClassificationModel._
+
+  // only called in copy()
+  def this(uid: String) = this(uid, 2, null)
+
+  private var trainingSummary: Option[XGBoostTrainingSummary] = None
+
+  /**
+   * Returns summary (e.g. train/test objective history) of model on the
+   * training set. An exception is thrown if no summary is available.
+   */
+  def summary: XGBoostTrainingSummary = trainingSummary.getOrElse {
+    throw new IllegalStateException("No training summary available for this XGBoostModel")
+  }
+
+  private[spark] def setSummary(summary: XGBoostTrainingSummary): this.type = {
+    trainingSummary = Some(summary)
+    this
+  }
+
+  // TODO: Make it public after we resolve performance issue
+  private def margin(features: Vector): Array[Float] = {
+    import DataUtils._
+    val dm = new DMatrix(scala.collection.Iterator(features.asXGB))
+    _booster.predict(data = dm, outPutMargin = true)(0)
+  }
+
+  private def probability(features: Vector): Array[Float] = {
+    import DataUtils._
+    val dm = new DMatrix(scala.collection.Iterator(features.asXGB))
+    _booster.predict(data = dm, outPutMargin = false)(0)
+  }
+
+  // The performance of single instance prediction is poor, need to be optimized later.
+  override def predict(features: Vector): Double = {
+    val _probability = probability(features)
+    val probabilityVec = if (numClasses == 2) {
+      Vectors.dense(Array(1.0 - _probability(0), _probability(0)))
+    } else {
+      Vectors.dense(_probability.map(_.toDouble))
+    }
+    probabilityVec.argmax
+  }
+
+  // Actually we don't use this function at all, to make it pass compiler check.
+  override def predictRaw(features: Vector): Vector = {
+    Vectors.dense(margin(features).map(_.toDouble))
```

Comment:
as all the single-instance-involved methods are slow but public, shall we throw exception just in case the user mistakenly call them?